### ISSUE TITLE
very slow initial run of query (unless explain is used first)

### ISSUE THREAD

**Issue by charlieschwabacher** [neo4j/neo4j#6615]
- Neo4j version: 2.3.0
- Operating system: osx 10.11.3
- API/Driver: cypher
### Steps to reproduce

starting from an empty DB, run:

```
CREATE (:aNode)-[:aEdge]->(:bNode)-[:bEdge]->(:cNode)-[:cEdge]->(:dNode)-[:dEdge]->(:eNode)-[:eEdge]->(:fNode)-[:fEdge]->(:gNode)-[:gEdge]->(:hNode)-[:hEdge]->(:iNode)-[:iEdge]->(:jNode)-[:jEdge]->(:kNode)-[:kEdge]->(:lNode)-[:lEdge]->(:mNode)-[:mEdge]->(:zNode)
```

and then run:

```
MATCH (a:aNode)-[:aEdge]->(b:bNode)-[:bEdge]->(c:cNode)-[:cEdge]->(d:dNode)-[:dEdge]->(e:eNode)-[:eEdge]->(f:fNode)-[:fEdge]->(g:gNode)-[:gEdge]->(h:hNode)-[:hEdge]->(i:iNode)-[:iEdge]->(j:jNode)-[:jEdge]->(k:kNode)-[:kEdge]->(l:lNode)-[:lEdge]->(m:mNode)-[:mEdge]->(z:zNode) RETURN a,b,c,d,e,f,g,h,i,j,k,l,m,z
```

OR

```
MATCH (a:aNode) MATCH (a)-[:aEdge]->(b:bNode) MATCH (b)-[:bEdge]->(c:cNode) MATCH (c)-[:cEdge]->(d:dNode) MATCH (d)-[:dEdge]->(e:eNode) MATCH (e)-[:eEdge]->(f:fNode) MATCH (f)-[:fEdge]->(g:gNode) MATCH (g)-[:gEdge]->(h:hNode) MATCH (h)-[:hEdge]->(i:iNode) MATCH (i)-[:iEdge]->(j:jNode) MATCH (j)-[:jEdge]->(k:kNode) MATCH (k)-[:kEdge]->(l:lNode) MATCH (l)-[:lEdge]->(m:mNode) MATCH (m)-[:mEdge]->(z:zNode) RETURN a,b,c,d,e,f,g,h,i,j,k,l,m,z
```
### Expected behavior

I would expect these queries to run quickly w/ only 14 nodes and 13 relationships in the db.
### Actual behavior

As I add nodes to my chain, this runs exponentially slower - taking about 70s w/ nodes `a` through `l`, and 370s w/ nodes `a` through `m` on a newish macbook pro.  I haven't waited for anything beyond that to complete.

After the very slow initial run, the same query will complete in around 10ms.  Interestingly, if I first run an `explain` on the query, the slow initial run can be entirely avoided.

I see the same behavior whether I run from the neo4j browser, or using cypher through node-neo4j.

**Comment by jexp** [comment]
@charlieschwabacher 
It is not the query that's slow in this case but the query planning.
I presume if you run the query prefixed with EXPLAIN it will take as long as well but then running it (even for the first time) will be fast.

Can you try the same on 2.3.2 and 3.0.0-M05 ?

Thanks

**Comment by jexp** [comment]
What is your use-case with this query?

**Comment by rationalism** [comment]
@jexp I've confirmed that planning these queries with EXPLAIN takes forever on my machine (Linux 2.3.2). Switching to the rule-based planner, by pre-pending CYPHER planner=rule, makes it fast (< 100 ms). 

The only time I've run into this kind of problem is when I used very large queries as a hack to get around latency issues - if, eg., a trip to the server takes 50 ms, it's faster to use one large query than ten small queries. But bundling lots of small queries into a single transaction has been working OK, so I haven't had to do that in a while.

**Comment by charlieschwabacher** [comment]
@jexp I didn't catch it before, but you are right - running the first explain is just as slow as actually running the query, and adding `CYPHER planner=rule` makes it fast.

I realize these are pretty atypical queries - I noticed the issue and came up w/ them while writing benchmarks to evaluate neo4j as the database for a GraphQL schema defined server framework.

**Comment by craigtaverner** [comment]
This issue was identified and fixed in January, with fixes going into 2.3 and 3.0 branches. Unfortunately it did not make the 2.3.2 release, so it will only come out in a future 2.3.3 (hopefully soon), but is available in the latest 3.0 M05 too. The details of the fix can be found in the PR description at https://github.com/neo4j/neo4j/pull/6269

But a short summary is that the query planning algorithm we use 'IDP' has a weakness for very long pattern expressions. Anything over length 6 or 7 starts taking exponentially longer to plan, due to combinatorial explosion. We investigated many options for reducing the search space without risking finding local minima (less perfect plans), and changed a few default settings in the IDP planner to allow it to find good solutions for very long expressions. We targeted up to length 30 within reasonable time (order of seconds).

In 2.3.2 and earlier, you can also work around this by setting planner=greedy. While greedy is not as good as IDP in most cases, it is better in this one. For 2.3.3 and later, the defaults should work well.
